armi/reactor/composites.py

```diff
@@ -525,6 +525,10 @@ def getChildren(self, deep=False, generationNum=1, includeMaterials=False):
         """Return the children of this object."""
         raise NotImplementedError
 
+    def iterChildrenWithFlags(self, typeSpec: TypeSpec, exactMatch=True):
+        """Produce an iterator of children that have given flags."""
+        raise NotImplementedError()
```

Comment:
Not sure I follow.  Why is this not implemented? It's `iterChildren` with a predicate, right?  Can that not be this high up the hierarchy for some reason?

Reply:
Oh excellent point. I'll move the `Composite.iterChildrenWithFlags` up to the `ArmiObject`

Reply:
Okay so `ArmiObject.iterChildren` is not defined and it maybe could be because `ArmiObject.getChildren` is defined

https://github.com/terrapower/armi/blob/c078cde00832c559586cbcc255356bac445a7a30/armi/reactor/composites.py#L524-L526

Since `Composite.iterChildrenWithFlags` relies on `Composite.__iter__` with `filter(lambda o: o.hasFlags(...), self)` (more or less) we could either
1. Implement `ArmiObject.iterChildrenWithFlags` as `ArmiObject.iterChildren(predicate=...)`, or
2. Enforce/assume that *every* subclass of `ArmiObject` is iterable and produces further `ArmiObject`s upon iteration (okay for non-leafs) and then implement `ArmiObject.iterChildrenWithFlags` as a filtered iteration over `self`

Both get us to the same place. My question is more from an object-oriented perspective: is it acceptable to assume/enforce that `ArmiObject`s should have `__iter__` implemented _somehow_?

Reply:
Opted to rely on the presence of `ArmiObject.iterChildren -> Iterator[ArmiObject]` which allows us to build a lot of iter/get functionality